### PR TITLE
[WIP] Add SQL Directives 

### PR DESCRIPTION
**Pull request by ani-sha**
- SQL directives for `relation` on table dependencies. 
- Need to handle the nested graphQL query results. (SQL Query executes, needs some handling with the GraphQL query)

### CODE REVIEW

**Comment by ani-sha** [comment]
FYI - Directives can be created for GraphQL `type`, but using them in `SelectionSet` in [Build SQL ](graphqlcrud-java/engine/src/main/java/io/graphqlcrud/SQLDataFetcher.java)  wasn't possible/accessible. So, used directives at the field level.

File: engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java [hunk 1]
```diff
@@ -56,24 +56,33 @@ public void testDataFetcher() throws Exception {
...
                     "}";
             String result3 = executeSQL(query3,graphQLSchema);
-            Assertions.assertEquals(result3,"SELECT SSN FROM PUBLIC.CUSTOMER WHERE SSN = 'CST01002'");
+            Assertions.assertEquals(result3,"SELECT PUBLIC.CUSTOMER.SSN FROM PUBLIC.CUSTOMER WHERE SSN = 'CST01002'");
 
+            String query4 = "{\n" +
+                    "  customers {\n" +
+                    "    accounts {\n" +
+                    "      SSN\n" +
+                    "    }\n" +
```
Comment (rareddy) [inline engine/src/test/java/io/graphqlcrud/SQLDataFetcherTest.java:81]: Here SSN is field in customer ( iknow it is also in accounts but just relation key), and accounts has their own field names for ex:
```
customers {
      SSN        
      accounts {
         TYPE,
          ACCOUNT_ID
       }
 }
```

**Comment by wtrocki** [comment]
> know I mentioned the description before, unless it is graphql specific let's not do this. Also, I would like to understand little more on directive vs comment, as directives are available everywhere pretty easily not sure that is case with description (maybe it is)

Directives have a lots of problems if you wish to process schema using them. If directive is used only to mark stuff on build annotation will be needed.

Generally in your case db dictates everything so there is no need to have directive or annotation (as we have all info about relations from database). Description can be used to inform users etc. and I think annotations here work better

**Comment by rareddy** [comment]
> Directives might have a lots of problems if we wish to process schema using them. If a directive is used only to mark elements on build time, the annotation will be easier

The description or annotation will force you to define a small DSL that will force you to write additional parsers etc. I feel directives are better to extend the metadata as GraphQL schema language allows them. However, I would like to know more about the issues they arise though. I feel they are designed to be used in this kind of situation.

> Generally in our case db dictates everything so there is no need to have directive or annotation (as we have all info about relations from database).

Yes, agree DB has the details, however, IMO the schema should be self-reliant about what it is exposing from DB. When we develop a top-down approach these annotations or directives will help define to drive DDL that can be created, keeps both models decoupled. It is a separate issue consider in each scenario when a value is not supplied through directive/annotation what the default behaviour going to be.

**Comment by wtrocki** [comment]
Ok. No problem to use directive. We need to define them on top of the schema in order for schema to be valid.

**Comment by ani-sha** [comment]
> Ok. No problem to use directive. We need to define them on top of the schema in order for schema to be valid.

It would be great if you could elaborate this as a small example.

**Comment by wtrocki** [comment]
https://www.graphql-java.com/documentation/v11/sdl-directives/

**Comment by rareddy** [comment]
thanks for that link @wtrocki  From the examples, the one troubling part (which is Ok IMO) is directives are both being used as design-time metadata and also runtime metadata. for example, using `sql` directive we are planning on using one can change values at runtime too that will end up in error. I have not encountered a scenario where we would need that but it could be useful sometimes to send a hint on query planning.

**Comment by ani-sha** [comment]
Basic query execution with directives. Going ahead to merge into master.

**Pull request merged**